=pro move ssl-config-akka over from ssl-config project

jrudolph opened the pull request:
Without further change it means that ssl-config-akka will be published
under the com.typesafe.akka organization and the akka versioning (which
probably makes sense).

Source code was copied over from the tree at commit
typesafehub/ssl-config@470fae7

See https://github.com/typesafehub/ssl-config/issues/47

patriknw commented on project/AkkaBuild.scala:
```diff
@@ -306,11 +306,17 @@ object AkkaBuild extends Build {
   )
 
+  lazy val sslConfigAkka = Project(
+    id = "ssl-config-akka",
```
I think the name should be akka-ssl-config

ktoso replied:
+1, sounds better indeed

ktoso reviewed (changes requested):
LGTM after artifact name change

jrudolph committed:
=pro merge ssl-config-akka from ssl-config project into akka-stream

This will fix the cyclic dependency issue between the ssl-config repo
and akka.

It would have been better if akka-stream would not require
these changes at all but com.typesafe.sslconfig.akka.AkkaSSLConfig is
part of the public interface of akka-stream's TLS stage. So, this
can only be fixed in the next major version.

Source code was copied over from the tree at commit
https://github.com/typesafehub/ssl-config/commit/470fae76f38dbbd4a42a4000675f66fa67e5f1da

See https://github.com/typesafehub/ssl-config/issues/47

jrudolph commented:
PLS BUILD

jrudolph commented:
Fixed the naming.

patriknw commented on akka-ssl-config/src/main/scala/com/typesafe/sslconfig/akka/AkkaSSLConfig.scala:
```diff
@@ -0,0 +1,211 @@
+    new AkkaSSLConfig(system, defaultSSLConfigSettings(system))
+
+  def defaultSSLConfigSettings(system: ActorSystem): SSLConfigSettings = {
+    val akkaOverrides = system.settings.config.getConfig("akka.ssl-config")
```
where is `akka.ssl-config` defined? above reference.conf has `ssl-config`

jrudolph replied:
IIRC I did not change anything. So, if there's a problem it would already exist in the existing module. I'll have a look...

jrudolph replied:
@ktoso do you know how this is supposed to work?

patriknw replied:
`getConfig("akka.ssl-config")` will throw an exception if it's not defined in some reference.conf, and to me it feels like that should be in the reference.conf of this library.

ktoso replied:
We should put an empty section that's akka.ssl-config into this reference as well.
In general it's set in akka-streams AFAIR, the only user of this lib

patriknw replied:
ok, I see that, but that assumption is not enough, since it should be possible to use without akka-stream

patriknw commented on akka-ssl-config/src/main/scala/com/typesafe/sslconfig/akka/AkkaSSLConfig.scala:
```diff
@@ -0,0 +1,211 @@
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package com.typesafe.sslconfig.akka
```
I think I'm missing the background for this library. So there is already something called ssl-config-akka published that is using exactly this package name?
I'm afraid of problems that will occur with transitive dependencies including both ssl-config-akka and this akka-ssl-config, and they are using same package/class names.

jrudolph replied:
Yes, it's the same module (and should become a replacement of the old one) so we should reconsider if renaming could be a problem.

Assuming that no one references this module manually, it probably wouldn't be a problem, because akka-stream will just point to the new module. Otherwise, if people  already depend on the old module explicitly, they might end up with to versions on their classpath if they also use akka-stream in the new version now depending on the new module.

So, keeping the old artifact coordinates could make sense. In this case, the module would need to be special-cased to also use the previous maven organization.

patriknw replied:
then it sounds like we could change the package name, to be on the safe side?

ktoso replied:
So the general idea was to apply to akka-specific types things in this lib, while ssl-config-core only can touch SSLContext / Params etc. AFAICS current state of lib does not really need akka specific types other than setting up the logger etc. This extension allows changing the settings before being passed into akka http programatically, in case you have more apps in the same jvm etc.

I agree with changing the package name here to perhaps com.lightbend

ktoso commented on akka-ssl-config/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,8 @@
+# Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
```
Lightbend

ktoso commented on akka-ssl-config/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,8 @@
+# ssl configuration
+ssl-config {
+
+  logger = "com.typesafe.sslconfig.akka.util.AkkaLoggerBridge"
```
This is correct, but needs documentation perhaps.
Generally speaking the "core ssl-config" wants to log things sometimes, and we need to provide it with the akka logger, thus this setting is for the "core ssl-config"

patriknw replied:
but it's bad practice to try to override someone else's reference.conf in a reference.conf, classpath ordering matters.

To me it looks like the intention was to use `akka.ssl-config.logger` here

ktoso commented on akka-ssl-config/src/main/scala/com/typesafe/sslconfig/akka/AkkaSSLConfig.scala:
```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
```
Lightbend

ktoso commented on project/OSGi.scala:
```diff
@@ -81,6 +82,12 @@ object OSGi {
+    Seq(
+    OsgiKeys.bundleSymbolicName := s"${organization.value}.sslconfig.akka",
+    OsgiKeys.exportPackage := Seq(s"com.typesafe.sslconfig.akka.*;version=${version.value}"),
+    OsgiKeys.requireBundle := Seq(s"""com.typesafe.sslconfig;bundle-version="${version.value}""""))
```
looks good

ktoso commented on akka-ssl-config/src/main/scala/com/typesafe/sslconfig/akka/util/AkkaLoggerBridge.scala:
```diff
@@ -0,0 +1,32 @@
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package com.typesafe.sslconfig.akka.util
```
yeah let's move it to `com.lightbend`, to avoid weird classpath issues I think, as Patrik suggested

ktoso commented on akka-ssl-config/src/main/scala/com/typesafe/sslconfig/akka/util/AkkaLoggerBridge.scala:
```diff
@@ -0,0 +1,32 @@
+import akka.event.Logging._
+import com.typesafe.sslconfig.util.{ LoggerFactory, NoDepsLogger }
+
+final class AkkaLoggerFactory(system: ActorSystem) extends LoggerFactory {
```
please add INTERNAL API marker?

ktoso reviewed (approved):
LGTM, let's change the package name though

jrudolph commented:
We discussed on gitter that given that the module currently has no real use for being standalone will be merged into akka-stream for now. It would have been better to prevent the dependency from akka-stream -> ssl-config-akka in the first place but we'll only be solve that knot in Akka 2.5 since the public TLS interface already depends on these classes.

ktoso commented on akka-stream/src/main/scala/com/typesafe/sslconfig/akka/AkkaSSLConfig.scala:
```diff
@@ -0,0 +1,212 @@
+import com.typesafe.sslconfig.util.LoggerFactory
+
+// TODO: remove again in 2.5.x, see https://github.com/akka/akka/issues/21753
+object AkkaSSLConfig extends ExtensionId[AkkaSSLConfig] with ExtensionIdProvider {
```
`private[akka]`?

ktoso replied:
Oh no wait... no maybe

ktoso commented:
LGTM

patriknw commented on akka-stream/src/main/resources/reference.conf:
```diff
@@ -96,3 +96,9 @@ akka {
+# folded in from former ssl-config-akka module
+ssl-config {
+  logger = "com.typesafe.sslconfig.akka.util.AkkaLoggerBridge"
+}
```
must be newline here, because of sbt-assembly (and similar) merging

jrudolph replied:
good point

jrudolph commented:
Updated to merge the files directly into akka-stream. I created a new ticket to fix for Akka 2.5 where should try again to get rid of the files in the first place: https://github.com/akka/akka/issues/21753
